### PR TITLE
First release

### PR DESCRIPTION
**Pull request by aodhan-domhnaill**
## `couch2pg` replacement

This package only copies records from CouchDB to Postgres

Features,

- **Append only** New doc revisions create a new record in the Postgres table
- **Streaming** New records are copied over in "realtime"
- **No code** This is just using LogStash and PostgREST together, so we don't own any code

Things you might worry about,

1. **We will have a bunch of duplicate records!!**

Yes. De-duping is easy in SQL and being append-only makes incremental builds much easier. For example, if we tried to update/delete/insert records into the initial table, then we need to worry about update/delete/insert operations in all downstream tables. It's much easier to simply do insert only, until the very last table and then filter

```
SELECT DISTINCT ON (_id)
  ...
FROM last_downstream_table
ORDER BY _id, _rev DESC
```

2. **What if LogStash crashes?**

Current setup is not stateless (unfortunately), so when LogStash comes back up, it will try to copy all the data into Postgres again. Postgres has a `UNIQUE (_id, _rev)` contraint on the root table, so it will simply fail to insert these records. This specific failure is ignored by Logstash.

3. **How will this fit in with current app dev practices?**

https://github.com/medic/cht-pipeline/pull/5 is trying to bridge the gap

### CODE REVIEW

File: logstash/pipeline/postgrest.conf [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    couchdb_changes {
+        always_reconnect => true
+        db => "${COUCHDB_DB}"
+        host => "${COUCHDB_HOST}"
+        username => "${COUCHDB_USER}"
+        password => "${COUCHDB_PASSWORD}"
+        keep_id => true
+        keep_revision => true
+        #sequence_path => "/var/couchdb/sequence"
+        #secure => true
```
Comment (aodhan-domhnaill) [inline logstash/pipeline/postgrest.conf:11]: Need to turn secure on and set sequence_path

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+version: '3.7'
+
+services:
+  logstash:
+    image: medic/logstash-couchdb:latest
```
Comment (aodhan-domhnaill) [inline docker-compose.yml:5]:
```suggestion
    image: medic/cht-logstash-couchdb:latest
```

**Comment by aodhan-domhnaill** [comment]
Docker image is pushing now

File: README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+POSTGRES_TABLE=couchdb
+POSTGRES_SCHEMA=v1
+
+COUCHDB_USER=<read access user>
+COUCHDB_PASSWORD=<password>
+COUCHDB_DB=<main database with all records>
+COUCHDB_HOST=<hostname for couchdb server>
+```
+
+2.  Create a table like in [`init-db.sh`](./init-db.sh)
```
Comment (njogz) [inline README.md:24]: Does this mean one should use `init-db.sh` to create the table or use it to get the schema

**Review by njogz — APPROVED** [review]
Looks good. I have left one question about an ambiguous step in the README file though

**Comment by aodhan-domhnaill** [comment]
Investigating a bug. The below change,

```
$ git diff
diff --git a/docker-compose.yml b/docker-compose.yml
index efd3caf..0c20318 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -11,7 +11,7 @@ services:
       - COUCHDB_DB=${COUCHDB_DB}
       - COUCHDB_HOST=${COUCHDB_HOST}
       - COUCHDB_PORT=${COUCHDB_PORT}
-      - COUCHDB_SEQ=/opt/couchdb/sequence_path.txt
+      - COUCHDB_SEQ=/tmp/couchdb/sequence_path.txt
       - COUCHDB_SECURE=${COUCHDB_SECURE:-true}
       - POSTGRES_TABLE=${POSTGRES_TABLE}
       - HTTP_ENDPOINT=postgrest:3000
```

Causes,

```
couchdb_1                   | [notice] 2022-02-03T12:29:22.763708Z nonode@nohost <0.9202.9> 2015c5e91a couchdb:5984 172.31.0.8 test GET /medic/_changes?feed=continuous&include_docs=true&since=&heartbeat=1000 400 ok 3
logstash_1                  | [2022-02-03T12:29:22,765][ERROR][logstash.inputs.couchdbchanges][main][6d96458f146fbed4fe572e56aa21cbb99a0bcb04637cfde4022a694c32493513] Unable to connect to database {:db=>"medic", :error=>"{:message=>\"Request error!\", :response_code=>\"400\"}"}
```

Which is confusing me.

**Comment by dianabarsan** [comment]
@aodhan-domhnaill 
It's because of your empty since. If you open that url in your browser (/medic/_changes?feed=continuous&include_docs=true&since=&heartbeat=1000) you'll see the full error. 
```
{"error":"bad_request","reason":"Malformed sequence supplied in 'since' parameter."}
```

`since` is either 0, or a valid seq.

**Comment by aodhan-domhnaill** [comment]
@nomulex I pushed a temporary fix. Can you try this in k8s and see if the sequence file is updating now? I moved it to `/tmp/couchdb/sequence_path.txt`

**Pull request merged**